libs/MeshKernel/src/CurvilinearGrid/CurvilinearGrid.cpp

```diff
@@ -1304,9 +1310,91 @@ std::vector<CurvilinearGrid::CurvilinearFaceNodeIndices> CurvilinearGrid::Comput
     return result;
 }
 
+std::set<CurvilinearGrid::CurvilinearEdge> CurvilinearGrid::ComputeBoundaryEdges(const CurvilinearGridNodeIndices& lowerLeft, const CurvilinearGridNodeIndices& upperRight) const
+{
+    std::vector<Point> result;
+    const auto facesIndices = ComputeFaceIndices(lowerLeft, upperRight);
+    std::set<CurvilinearEdge> boundaryEdges;
+
+    for (const auto& faceIndices : facesIndices)
+    {
+        if (!std::ranges::all_of(faceIndices, [this](const CurvilinearGridNodeIndices& curviNode)
+                                 { return GetNode(curviNode.m_n, curviNode.m_m).IsValid(); }))
+        {
+            continue;
+        }
+
+        const auto numFaceNodes = static_cast<UInt>(faceIndices.size());
+        for (UInt i = 0u; i < numFaceNodes; ++i)
+        {
+            const auto firstCurvilinearNodeIndex = faceIndices[i];
+            const auto nextIndex = NextCircularForwardIndex(i, numFaceNodes);
+            const auto secondCurvilinearNodeIndex = faceIndices[nextIndex];
+
+            const auto edge = std::make_pair(std::min(firstCurvilinearNodeIndex, secondCurvilinearNodeIndex),
+                                             std::max(firstCurvilinearNodeIndex, secondCurvilinearNodeIndex));
+
+            const auto it = boundaryEdges.find(edge);
+            if (it != boundaryEdges.end())
+            {
+                boundaryEdges.erase(it);
+            }
+            else
+            {
+                boundaryEdges.insert(edge);
+            }
+        }
+    }
+    return boundaryEdges;
+}
+
+std::vector<meshkernel::Point> CurvilinearGrid::ComputeBoundaryToPolygon(const CurvilinearGridNodeIndices& lowerLeft, const CurvilinearGridNodeIndices& upperRight) const
+{
+    std::vector<Point> result;
+
+    auto boundaryEdges = ComputeBoundaryEdges(lowerLeft, upperRight);
+    if (boundaryEdges.empty())
+    {
+        return result;
+    }
+
+    auto currentEdge = boundaryEdges.begin();
```

Comment:
comment: start with smallest first node

Reply:
it does not have to be

Reply:
you're using a set, that's ordered, so you'll sort by std::pair of vertex indices which compares first, then second, I believe. If you don't care you could use std::unordered_set which may be faster to build, but starting with the smallest has the advantage that you know you are on the outer boundary of the polygon, not one of the holes.